lib/github/markups.rb

```diff
@@ -52,3 +52,5 @@
 )
 
 command(::GitHub::Markups::MARKUP_POD, :pod2html, /pod/, ["Pod"], "pod")
+
+command(::GitHub::Markups::MARKUP_MANPAGE, :man2html, /1|2|3|4|5|6|7|8|man|mdoc/, ["Roff"], "manpage")
```

Comment:
This did not work until I referred to the language as "Roff" as linguist does here: https://github.com/github/linguist/blob/master/lib/linguist/languages.yml#L3994

We're probably going to want to update our regex to be consistent with that list. We might also want a PR to linguist to add '.mdoc' to the list of extensions, since we're using mandoc to render and it would be nice of us to support their preferred extension. I'm not sure if that's a separate language / sub language / same syntax highlighting or different from a linguist perspective; from a cursory glance, linguist seems to have no existing knowledge of it.

Reply:
`/man|1(in|[mx])?|2|3(in|qt|[mx])?|[4-9]|l|m[es]|ms|n|nr|rno|roff|tmac/` ought to work

Reply:
ia ia regex ftagn

Reply:
Yes, you should update regex to support also `.tmac`, `.roff` ideally also `.1.in`, `.2.in`, `.3.in`, ... , `.8.in` as more manpages contains autoconf variables and therefore are in git repository stored as file with extension `.in`.

Reply:
There's a myriad of section group suffices not being accounted for here. Solaris-based systems have a [particularly extensive](https://github.com/illumos/illumos-gate/tree/master/usr/src/man) range of group suffices.

Reply:
I'm jumping in part-way through, but is the aim to spot man-page source in source trees by file suffix alone? That won't match man/foo that's later installed by a makefile as man3/foo.3x or whatever the build system decides is the local convention.

For suffixes, my Arch Linux machine happens to have

    0p
    1 1m 1p 1perl 1ssl 1x
    2
    3 3am 3bsd 3guile 3lua 3ocaml 3p 3pcap 3perl 3pm 3python 3r
        3sharp 3ssl 3t 3tcl 3tiff 3x
    4
    5 5ssl
    6
    7 7ssl
    8
    n

In a source tree, these might be the same suffix, or the suffix `.man` or `.tr`, and also have `.in` appended, or not have a suffix. Is looking at the file's contents an option?

Reply:
Regarding languages and terminology:
_roff(7)_ is a general-purpose typesetting language that you can use to typeset anything - books, letters, posters, whatever.  It is purely presentational.  If something is written in roff(7), that neither implies that it is a manual page, nor that mandoc can handle it.
_mdoc(7)_ is a markup language specifically developed for manual pages by Cynthia Livingston at Berkeley in 1989/90.  While it was originally developed as a macro set on top of roff(7), it is a language in its own right in so far as a well-formed manual page will only use mdoc macros and will not directly use any low-level roff requests.  It is also fundamentally different from roff(7) in so far as it almost exclusively marks up semantic functions (this is a function name, a command line option, and argument placeholder to be replaced by whatever the user wants, and so on) and very rarely resorts to presentational statements like "print this in boldface" - just like HTML.
_man(7)_ is the original, historical markup language used by Bell Labs for the AT&T Unix Version 7 manual pages in 1979.  Calling it a "language" is a bit of a stretch because it really only provides a small, incomplete set of macros on top of roff(7), so you cannot really write anything in pure man(7), you always have to mix in low-level roff requests in between.  Besides, it renders poorly to HTML because it is purely presentational, so you get lots of unqualified "b" and "i" tags from it.  Strangely, the Linux Manpages Project still uses it.
What you really want to recognize is mdoc(7) and man(7) files, *not* roff(7) files at large.  That is how all manual pages are written and also what mandoc can handle.

That said, file extensions are an extremely poor way of identifying manual pages files, especially in source trees - it becomes a bit better once the files are installed, but only a bit.  If you have a choice, you want to ignore the file extension altogether and instead look at the *content* of the file.  If the first line matching the ERE /^([^.]|\.[A-Za-z])/ starts with exactly ".Dd" or ".TH", it is a manual page.  Otherwise, it is not.

If you are somehow forced to look at the file extension, you have to expect some errors in both directions.  The best algorithm i can think of goes as follows:
1. Strip trailing ".in", if it is present.  ==  sed '/\.in$//'
2. Strip everything up to and including the last dot.  == sed '/.*\.//'
3. If the result is exactly "man", exactly "mdoc", exactly "n", or starts with a single *non-zero* digit (1-9), optionally followed by one or more letters, it is (probably) a manual page.

You definitely don't want files with names like
1. *.roff -- wrong language, likely not a manual page, unless the content starts with .Dd or .TH
2. *.tmac -- not a manual page, but a troff macro definition file
3. *.me -- wrong language, that's the "Eric Allman" macro set, not a manual page
4. *.ms -- wrong language, that's the Unix v6 Bell Labs macro set, not a manual page
5. *.0 -- those are preformatted manual pages, essentially *.txt files; consequently, they rarely occur in source trees, but still, manual page formatters cannot handle them, you simply read them with a pager like more(1) or less(1)

I have never seen *.nr, *.rno, and *.tr, no idea what those are, likely not manual pages.

Reply:
> If the first line matching the ERE /^([^.]|\.[A-Za-z])/ starts with exactly ".Dd" or ".TH", it is a manual page. Otherwise, it is not.

That is not truth. Lot of manpage files on Linux starts with comments, including those from Linux manpages project. Below are some examples of the first lines of manpages in my system:
```
.\" t
'\" t -*- coding: UTF-8 -*-
.\"*
'\" e
'\" et
'\" t
\t
\" Copyright
.\"/* Copyright
.so man1/pdftex.1
.de EX
.ds PX \s-1POSIX\s+1
.EQ
```

Reply:
He means a *well-formed* manual-page. 😉 As in, one that can be reasonably assumed to be `mdoc` or `man` markup, irrespective of file extension. And the expression itself should be:

~~~regexp
/^[.'][ \t]*[A-Za-z]{2}/m
~~~

Or to be more permissive about macro/request names:

~~~regexp
/^[.'][ \t]*[^\s\\]+/m
~~~

> Below are some examples of the first lines of manpages in my system:

Some of those are malformed, BTW:
* `\t` does nothing unless read in copy mode (e.g., used in a diversion or string/macro definition), where it means a horizontal tab (which would be meaningless anyway).
* `.\" t` should be `'\" t` instead; this is simply an [idiom](https://www.gnu.org/software/groff/manual/html_node/Preprocessors-in-man-pages.html) to instruct pagers to use `tbl` or `eqn` for preprocessing input (`'\" e` meaning `eqn`, `'\" t` meaning `tbl`, and `et` or `te` meaning both).
* `\" Copyright` is an empty input line which ends in a comment; the author probably meant to use `.\" Copyright` or `'\" Copyright` instead.

Others are correct but unlikely to occur in `mdoc` documents. An exception might be `.so man?/*`, which essentially means "Load this file instead". It's a rough equivalent of a symlink in the manpage world.

Reply:
Replying to @ischwarze:
> If you have a choice, you want to ignore the file extension altogether
> and instead look at the content of the file.

I think they go by extension, and then have code that can discern
between differing languages that happen to share extension by inspecting
the start of the file.  Presumably, that code can also say it was
unknown after all, e.g. a `.tr` that turned out not to be troff.

> If the first line matching the ERE /^([^.]|\.[A-Za-z])/ starts with
> exactly ".Dd" or ".TH", it is a manual page. Otherwise, it is not.

Backing this up for a man page, as distinct from a mandoc page...
man(7) and man-pages(7) from package man-pages 4.16-2 on Arch Linux,
http://man7.org/linux/man-pages/index.html upstream, re-iterates this.

    The first command in a man page (after comment lines, that is, lines
    that start with .\") should be

	.TH title section date source manual

That's because it's a long-standing convention and programs that don't
really understand troff still parse the start of man pages for this one
line.  A man page that doesn't do this is buggy and can be ignored;  it
will help highlight the problem.

>     Strip trailing ".in", if it is present. == sed '/\.in$//'
>     Strip everything up to and including the last dot. == sed
>     '/.*\.//'

This misses that there must be a dot.

>     If the result is exactly "man", exactly "mdoc", exactly "n", or
>     starts with a single non-zero digit (1-9), optionally followed by
>     one or more letters, it is (probably) a manual page.

This excludes man pages in section 0, which is common.

>     *.roff -- wrong language, likely not a manual page, unless the
>     content starts with .Dd or .TH

Even then, I'd say it's wrong and can be ignored.  A man/mandoc page
would normally have a more specific suffix.

>     *.0 -- those are preformatted manual pages, essentially *.txt
>     files; consequently, they rarely occur in source trees, but still,
>     manual page formatters cannot handle them, you simply read them
>     with a pager like more(1) or less(1)

No, that might be true for OpenBSD, but not here on Linux, e.g.
tar.h(0p) is POSIX's header file.

> I have never seen *.nr, *.rno, and *.tr, no idea what those are,

First and last are nroff and troff source.  They're quite widely used,
e.g. a makefile rule will be from .tr to .ps.  Never seen .rno.

> likely not manual pages.

Agreed.

Reply:
@pali wrote:
> > If the first line matching the ERE /^([^.]|.[A-Za-z])/ starts with
> > exactly ".Dd" or ".TH", it is a manual page. Otherwise, it is not.
>
> That is not truth. Lot of manpage files on Linux starts with comments

But that's not what Ingo said.  He said the first line that matches that
regexp, not the first line in the file.  That said, his regexp doesn't
consider the other kind of troff command starting with a `'`.

> including those from Linux manpages project. Below are some examples
> of the first lines of manpages in my system:

With a better regexp, Ingo please check it's OK for mandoc, I find these
buggy man pages from a selection here.

    $ perl -n0777e '/\A(?:[.'\'']\s*\\".*\n)*(.*)/; $1 =~ /^\.(?:TH|Dd)/ or print "$ARGV\t$1\n"' *
    ab.1	.de Sh \" Subsection
    aclocal.1	.so man1/aclocal-1.15.1
    addr2line.1	.de Sp \" Vertical space (when we can't use .PP)
    alsa_out.1	.so man1/alsa_in.1
    anytopnm.1	\
    apropos.1	.pc
    apxs.1	.de Sh \" Subsection
    ar.1	.de Sp \" Vertical space (when we can't use .PP)
    aria2c.1	.
    arpaname.1	.hy 0
    as.1	.de Sp \" Vertical space (when we can't use .PP)
    asciitopgm.1	\
    asn1parse.1ssl	.de Sp \" Vertical space (when we can't use .PP)
    atktopbm.1	\
    automake.1	.so man1/automake-1.15.1

The `.so` are sourcing another file because the man page is known by
multiple names.  I think these should be omitted rather than formatted.

The others are bugs, typically from a small number of foo-to-man
converters.